// @craftjs/utils/.rollup.cache/Users/prevwong/Documents/github/craft.js/packages/utils/lib/EventHandlers/DerivedEventHandlers.js
import { EventHandlers } from './EventHandlers';
import { EventHandlerUpdates } from './interfaces';
// Creates EventHandlers that depends on another EventHandlers instance
// This lets us to easily create new connectors that composites of the parent EventHandlers instance
export class DerivedEventHandlers extends EventHandlers {
    constructor(derived, options) {
        super(options);
        this.derived = derived;
        this.options = options;
        // Automatically disable/enable depending on the parent handlers
        this.unsubscribeParentHandlerListener = this.derived.listen((msg) => {
            switch (msg) {
                case EventHandlerUpdates.HandlerEnabled: {
                    return this.enable();
                }
                case EventHandlerUpdates.HandlerDisabled: {
                    return this.disable();
                }
                default: {
                    return;
                }
            }
        });
    }
    // A method to easily inherit parent connectors
    inherit(cb) {
        return this.createProxyHandlers(this.derived, cb);
    }
    cleanup() {
        super.cleanup();
        this.unsubscribeParentHandlerListener();
    }
}
